docs/blockentities/container.md

```diff
@@ -0,0 +1,250 @@
+
+```java
+public class MyBlockEntity extends BaseContainerBlockEntity {
+    // The container size. This can of course be any value you want.
+    public static final int SIZE = 9;
+    // Our item stack list. This is not final due to #setItems existing.
+    private NonNullList<ItemStack> items = NonNullList.withSize(SIZE, ItemStack.EMPTY);
+
+    // The constructor, like before.
+    public MyBlockEntity(BlockPos pos, BlockState blockState) {
+        super(MY_BLOCK_ENTITY.get(), pos, blockState);
+    }
+
+    // The container size, like before.
+    @Override
+    public int getContainerSize() {
+        return SIZE;
+    }
+
+    // The getter for our item stack list.
+    @Override
+    protected NonNullList<ItemStack> getItems() {
+        return items;
+    }
+
+    // The setter for our item stack list.
+    @Override
+    protected void setItems(NonNullList<ItemStack> items) {
+        this.items = items;
+    }
+
+    // The display name of the menu. Don't forget to add a translation!
+    @Override
+    protected Component getDefaultName() {
+        return Component.translatable("container.examplemod.myblockentity");
+    }
+
+    // The menu to create from this container. See below for what to return here.
+    @Override
+    protected AbstractContainerMenu createMenu(int containerId, Inventory inventory) {
+        return null;
+    }
+}
+```
+
+Keep in mind that this class is a `BlockEntity` and a `Container` at the same time. This means that you can use the class as a supertype for your block entity to get a functioning block entity with a pre-implemented container.
+
+## Using Containers
+
+Now that we have created containers, let's use them!
+
+Since there is a considerable overlap between `Container`s and `BlockEntity`s, containers are best retrieved by casting the block entity to `Container` if possible:
+
+```java
+if (blockEntity instanceof Container container) {
+    // do something with the container
+}
+```
+
+The container can then use the methods we mentioned before, for example:
+
+```java
+// Get the first item in the container.
+ItemStack stack = container.getItem(0);
+
+// Set the first item in the container to dirt.
+container.setItem(0, new ItemStack(Items.DIRT));
+
+// Removes a quantity of (up to) 16 from the third slot.
+container.removeItem(2, 16);
+```
+
+:::warning
+A container may throw an exception if trying to access a slot that is beyond its container size.
+:::
+
+## `Container`s on `ItemStack`s
+
+Until now, we mainly discussed `Container`s on `BlockEntity`s. However, they can also be applied to [`ItemStack`s][itemstack] using the `minecraft:container` [data component][datacomponent]:
+
+```java
+// We use SimpleContainer as the superclass here so we don't have to reimplement the item handling logic ourselves.
+// You may of course use a different base implementation of Container (or implement Container yourself) if needed.
+public class MyBackpackContainer extends SimpleContainer {
```

Comment:
This is very wonky as because the component maintains a copy of the stacks on update, this will not reflect any changes from someone else operating on the component at the same time. My immediate thought is the bundle, as that would be interacting with the data while in a container, which could be accessed by multiple people.

Not really sure what would be the best way to handle this, unless you are also getting the items from the container contents based on the provided slot.

Reply:
I added a note that race conditions can happen this way, and just assumed that only one player can access the backpack on this example.

Reply:
That seems fine to me. If a better solution comes up later, we can always replace it.